### DISCUSSION TITLE
Issue with io/core2

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#2043] [General]
The `core2` dependency (primarily needed because of our `Encodable`/`Decodable` traits usage of `io::Write` and `io::Error`) is problematic. There have been many discussions spread out all over the place. This is an attempt to bring them all together.

### Problems:

- The `core2` dependency broke our MSRV.
- We need the `core2` dependency because of the use of the `io::Write` trait in `Encodable`.
- `io::Error` does not, and cannot, implement `PartialEq`, this propagates through other error types.
- Writing to hashes is infallible and our code is littered with `.expect("writing to hashes can't fail")`
- Writing to in memory buffers only fails if re-allocation fails i.e., out of memory.
- `rust-bitcoin` requires an allocator but we mix IO and allocation up under the "no-std" feature

### Possible solutions/improvements:

- Add associated error type to `Decodable`, then decoders that do not return `io::Error` can use a different type.
- Add trait method `consensus_encode_infallible`

### Related links

- https://github.com/rust-bitcoin/rust-bitcoin/issues/2034
- https://github.com/rust-bitcoin/rust-bitcoin/issues/758
- https://github.com/rust-bitcoin/rust-bitcoin/issues/1251
- https://github.com/rust-bitcoin/rust-bitcoin/pull/1716

**Comment by tcharding** [comment]
FTR I do not fully grok everything that is going on with all the PRs/issues/discussion around this topic. I can work on this but I have no clue where to start, where we are trying to get to, or how to make progress.

**Comment by tcharding** [comment]
cc @TheBlueMatt because he raised #2034 and @LLfourn because he represents "bdk wants no-std feature gone".

**Comment by TheBlueMatt** [comment]
Yea, so I do wonder if we just go our own way on write traits - build a simple write trait with an associated type for the error, then hashes can make the error type `!` and its compiler-enforced they can't panic. Then, if the `std` feature is enabled we also impl our custom trait for all `std::io::Write`, which should be kinda sorta easy?

**Reply by apoelstra** [reply]
This suggestion has come up many times -- the obvious reason not to do this is that if we make our own `Write` trait then we'll be incompatible with everyone using core2. And if we convince people to use *our* trait, then they'll want a matching `Read` and `Seek` and `Error` etc etc etc and before we know it we're basically maintaining a copy of core2.

core2 itself is annoying enough that this still might be the way to go ... but it is pretty high-cost in my view.

**Reply by TheBlueMatt** [reply]
I'm not convinced the cost is that high - almost nothing uses `core2` except rust-bitcoin ecosystem crates, so I'm kinda dubious there is a single user who would get bitten by this.

**Reply by Kixunil** [reply]
We can still implement those traits on `std`/`core2`/... types (although we really should use weak dependencies for that) and it'll work seamlessly. Also we can provide bridges. But this is what `lgio` already does.

**Comment by RCasatta** [comment]
https://docs.rs/embedded-io/latest/embedded_io/ could probably be an interesting replacement but unfortunately its MSRV is 1.60

**Reply by tcharding** [reply]
We could fork it and rip out the `defmt` stuff (some sort of logging framework for embedded devices, that uses edition 2021)? Is that better or worse than forking `core2`, I don't know?

**Reply by apoelstra** [reply]
If we're going to fork something we might as well just write our own.

**Reply by RCasatta** [reply]
It seems the required changes for them to have MSRV 1.48.0 are minimal
https://github.com/rust-embedded/embedded-hal/pull/500 
let's see what they say

**Reply by TheBlueMatt** [reply]
I admit I'm not really convinced as long as they have any third party dependencies.

More generally, I don't understand taking on a dependency for anything so trivial unless there's real value. There was some discussion of it being to better integrate with other rust crates that use the same crate for IO, but it doesn't seem like there is some "rust ecosystem standard" here, so I really don't see a reason to.

**Reply by apoelstra** [reply]
Yeah, I'm coming around to @TheBlueMatt's view.

Especially as the rust-bitcoin ecosystem is now large enough that we could probably define our own "standard" for stuff like this which would be sufficient for all the HWW/wasm people using our stuff.

**Reply by RCasatta** [reply]
> I admit I'm not really convinced as long as they have any third party dependencies.

embedded-io has no mandatory deps

>  but it doesn't seem like there is some "rust ecosystem standard" here, so I really don't see a reason to.

This crate is maintained by an official rust working group, I think it has more chance to get more traction than ours or others.


In any case, it probably doesn't matter looking at one of the maintaner reaction I don't think they are willing to have MSRV 1.48.0 https://github.com/rust-embedded/embedded-hal/pull/500#issuecomment-1707145081

**Reply by TheBlueMatt** [reply]
> This crate is maintained by an official rust working group, I think it has more chance to get more traction than ours or others.

Ah, that's nice, yea. That said, if std moves to (finally...) taking some of these traits into `core` upstream then it'll kill people using this crate again, no?

**Reply by TheBlueMatt** [reply]
We can also revisit this in a year if/when embedded-hal's io crate has some traction.

**Comment by TheBlueMatt** [comment]
Spent some time working on implementing our own `io` lib on the plane today (cause LDK users are feeling this pain rather acutely - https://github.com/lightningdevkit/rust-lightning/pull/2560#issuecomment-1712356018), I think there's two designs here:

(1) Create our own Read/Write traits and (optionally) impl them for all std::io::Read/Write. This is ideal in theory I think - we get our own traits, no core2 stuff, no third party dependencies, and anyone who wants to use rust-bitcoin with std traits can simply pass in std traits where relevant. We can even have a feature to impl it all for std. Easy-peasy.

Sadly, this isn't trivial because `std::io::Read/Write` are already implemented for a blanket `&mut std::io::Read/Write`. We can't mirror this API while implementing our trait for all `std::io::Read/Write` without having conflicting implementations. So we'd have to skip that half of the impl and then not rely on this impl in rust-bitcoin. This isn't too hard for the Write side of things, a few `&mut`s here or there and it all works fine, but its much more annoying for `Read`.

`Read`'s normal API also relies on this blanket impl for various bits - eg `take` takes `self` by ownership and returns a wrap
er struct. Without the `&mut` blanket impl this would mean you can't ever wrap an `&mut Read` in a `Take`, which is incredibly
hard to work with. Instead, we could diverge further from the `std` API by making `Take` hold an `&mut Read`, and tweaking the
API here or there. Its mostly twelve of one half dozen of the other, but we'd have to start diverging in a few ways.

(2) Basically go the `core2` route, either re-exporting `std::io` types or our own identical types depending on a feature flag.
 This sadly has the same drawbacks as `core2` (except we can maintain it and not have it break our MSRV) - namely that with and
 without the `std` feature we have two logically different APIs, and mixing and matching isn't possible across different (transitive) dependencies.

**Reply by apoelstra** [reply]
Can we do both? Create our own Read/Write traits, implement these alongside std::io::Read/Write, and expose them unconditionally? Then std-only users can use the std traits and people who want to be more careful can use our traits (and deal with the inconvenience of missing blanket impls etc)

**Reply by TheBlueMatt** [reply]
Yea that's fine, it just means we diverge a bit from the std traits. They won't be full drop-in replacements but not different in any material ways. If that's the preference (option 1, above), I can finish up the pr and open it this week.

**Reply by TheBlueMatt** [reply]
Or, maybe I'm not understanding quite what you're describing, I don't we how we do both.

**Reply by apoelstra** [reply]
@TheBlueMatt for everything that we currently use `io::Write`, we also use our own `Write` trait in parallel.

**Reply by TheBlueMatt** [reply]
I think if we're gonna do that we should just go with option (1)? If we write our own trait and blanket-impl it for `std` then we get that "for free" but don't have to write the code in `rust-bitcoin` twice?

**Reply by LLFourn** [reply]
+1 for keeping it simple by not trying to mirror std traits perfectly (just the main required methods of each trait). People who are using `no_std` can cope with not having `Take` etc. They have to cope with that anyway. Doing a blanket impl for std::io::* seems like it will work. If it is inconvenient for some reason you can always make a macro that emits both impls. I deal in these kinds of macros. Here's one I just cooked up that supports type parameters and type constraints (not tested or compiling):

```rust
macro_rules! impl_read {
    (impl$(<$($tpl:ident  $(: $tcl:ident)?),*>)? Read for $type:path {
        fn read(&mut self, $buf:ident : &mut [u8]) ->  Result<usize> $block:block
    }) => {
        #[cfg(feature = "std")]
        impl $(<$($tpl $(:$tcl)?),*>)?  std::io::Read for $type {
            fn read(&mut self, $buf: &mut [u8]) -> Result<usize> {
                use std::io::Read as Read;
                $block
            }
        }

        impl $(<$($tpl $(:$tcl)?),*>)? crate::io::Read for $type {
            fn read(&mut self, $buf: &mut [u8]) -> Result<usize> {
                use crate::io::Read as Read;
                $block
            }
        }
    };
}

// impls both Reads for Thing where .inner implements both read traits
impl_read! {
    impl Read for Thing {
        fn read(buf: &mut [u8]) -> Result<usize> {
            Read::read(&mut thingy.inner, buf)
        }
    }
}
```

**Reply by TheBlueMatt** [reply]
I've got the base traits in place, though still working on getting it all cleaned up, adding errors, and working no-std + std - https://github.com/TheBlueMatt/rust-bitcoin/commits/2023-09-io Let me know if the read/write traits specifically and the changes in rust-bitcoin are reasonable.

**Reply by apoelstra** [reply]
The traits all look good to me. What is the plan for the `HashEngine` types which currently implement `io::Write`?

**Reply by TheBlueMatt** [reply]
For now I added a macro which does the correct (broadest) impl based on feature flags, see https://github.com/rust-bitcoin/rust-bitcoin/pull/2066

**Comment by apoelstra** [comment]
Ah I understand @TheBlueMatt. Sounds good to me!

Would be (begrudgingly) ok with a macro but I think the blanket impl will work.

**Reply by apoelstra** [reply]
It looks like the only place we actually impl `io::Write` is on the hashengines. And we don't impl `io::Read` anywhere. If I understand you correctly, we would preserve these impls but then also do a blanket `impl<T: io::Write> bitcoin::Write for T`?
